### ISSUE TITLE
printf format specifier incorrectly identified in multiline strings

### ISSUE THREAD

**Issue by smoothdeveloper** [dotnet/fsharp#2729]
Having this in a .fsx file:

```fsharp
sprintf
                "%A\n\
                 Dependencies -\n\
                 %s\n\
                 Source - %A\n\
                 Install Settings\n\
                 %A" "" "" "" ""
```

In VS2015 there is no error reported, but VS2017 wrongly reports this error:

> Type mismatch. Expecting a
>    ''a -> string -> 'b -> 'c -> 'd'    
>but given a
>    ''a -> string -> 'b -> string'    
>The type ''a -> 'b' does not match the type 'string'	

The snippet executes fine in FSI from VS in both cases.

**Workaround** - use  ``"""..."""`` multiline strings for the format specifier, which work correctly. These don't require ``\`` at the end of line

**Comment by dsyme** [comment]
@smoothdeveloper Are you sure?  I just compiled it using fsc.exe with VS2017 OK

**Comment by 0x53A** [comment]
I think this is a design time only issue:

```F#

// Learn more about F# at http://fsharp.org
// See the 'F# Tutorial' project for more help.

[<EntryPoint>]
let main argv =
    let x =
        sprintf
            "%A\n\
                Dependencies -\n\
                %s\n\
                Source - %A\n\
                Install Settings\n\
                %A"         "" "" "" ""
    printfn "%A" argv
    0 // return an integer exit code

```

![image](https://cloud.githubusercontent.com/assets/4236651/24471786/838fca7a-14c3-11e7-9342-534d7d6b7045.png)

Note how it doesn't color the last %A.

But it compiles fine:

> ========== Rebuild All: 1 succeeded, 0 failed, 0 skipped ==========

**Comment by forki** [comment]
Lol that's taken from paket source right? I saw that as well and had it on my ever growing investigation list.

**Comment by smoothdeveloper** [comment]
@dsyme this is actually in the tooling only, I indicated that it executes fine in FSI, but somehow the tooling (in VS2017 nightly) is behaving another way.

@forki yes, taken from Paket :smile:

**Comment by cloudRoutine** [comment]
It's the last trailing backslash that causes this, I've actually run into this error a ton of times.

![](http://i.imgur.com/w6u2OCk.png)

![](http://i.imgur.com/k4hNNPu.png)

**Comment by dsyme** [comment]
It's very odd, since the same text is being processed by the compiler and the language service.  Perhaps there's a `\r\n` v `\n` difference in the text sent to the language service and that processed by the compiler

**Comment by dsyme** [comment]
The problem here is that https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/CheckFormatStrings.fs#L52 is not taking into acount that simple ``"abc"`` strings in F# can use ``\`` at the end-of-line.  This means that the reconstructed format string is incorrect, with dangling ``\`` characters

The fix is probably a simple hack to take that kind of string into account when regenerating the format string.

TBH I don't like that code at all - it shouldn't be re-parsing the original source, and it probably shouldn't be needed at all.  It would be good to dig out the case that added it (in the FCS repository no doubt) and look if there is a simpler way

**Comment by abelbraaksma** [comment]
> I think this is a design time only issue:

@0x53A, no it isn't. There are actually three distinct cases with multilining and backslashes with PrintFormat specifiers that go wrong:

0. FS0020 warning in some specific cases, only in editor, not by compiler
0. FS0001 error, only in editor, not by compiler
0. FS3118 and FS0010 errors, both in editor *and in compiler*.

The latter being the most severe. The referenced issue #3380 gives a reproducible example for each of them. Here I'll just repeat the most severe one, as @dsyme  marked this as duplicate, but it hasn't been mentioned in this thread yet:

This won't compile:
````
module Multi =
    let foo = "all of us"

    // throws FS3118 and FS0010, each error twice
    let multiLineString = 
        sprintf "hello \
        world \
        here \
        we are\
        : %s" foo
````

**Comment by vasily-kirichenko** [comment]
The bug is still in 15.3.5

![1](https://user-images.githubusercontent.com/873919/30637725-d9a1adda-9e01-11e7-8cb6-1888a5bb3da6.gif)

**Comment by dsyme** [comment]
Keeping #7734 as the canonical bug - both are fixed by #10003
